README.md
# preact-leaflet [![codecov](https://codecov.io/gh/kontrollanten/preact-leaflet/branch/master/graph/badge.svg)](https://codecov.io/gh/kontrollanten/preact-leaflet) [![CircleCI](https://circleci.com/gh/kontrollanten/preact-leaflet.svg?style=svg)](https://circleci.com/gh/kontrollanten/preact-leaflet)

Supporting following Leaflet components:
* [Map](https://leafletjs.com/reference-1.4.0.html#map-example)
* [Marker](https://leafletjs.com/reference-1.4.0.html#marker)
* [MarkerCluster](https://github.com/Leaflet/Leaflet.markercluster)
* [Polyline](https://leafletjs.com/reference-1.4.0.html#polyline)
* [TileLayer](https://leafletjs.com/reference-1.4.0.html#tilelayer)
* [ZoomControl](https://leafletjs.com/reference-1.4.0.html#control-zoom)

## Example

```
import { Map, TileLayer } from 'preact-leaflet';

<Map center={[63.83919, 20.15069]} style={{ height: '100%' }} zoom={10}>
  <TileLayer url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
</Map>
```

For more advanced examples, look into the [e2e App](e2e/App.js) and it's [published version](https://preact-leaflet.netlify.com/).

## API
This is just a thin wrapper, checkout [Leaflet API](https://leafletjs.com/reference-1.4.0.html) for usage.

### Options
All props, except for `on` prefixed, are passed to Leaflet.

### Event listeners
Props prefixed with `on` are handled as event listeners, e.g. `onZoom` will listen for the `zoom` event.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.1.0 (2019-12-10)


### Features

* **Map:** add support for all leaflet Map options ([90a6e28](https://github.com/kontrollanten/preact-leaflet/commit/90a6e283286de540480b4cb32a53b897138274b0))
* **Map:** add support for event listeners ([f211aaf](https://github.com/kontrollanten/preact-leaflet/commit/f211aafa09aa3e199512d901c73b2379a24f1aa0))
* **Map:** add support for passing bounds prop ([e6e3f9e](https://github.com/kontrollanten/preact-leaflet/commit/e6e3f9efa1d191fae521cc78e0e8326b05665399))
* **Map:** add support to update center position ([7395021](https://github.com/kontrollanten/preact-leaflet/commit/7395021ea2d686baea398ae63419853963114a4a))
* **Map:** add support to update zoom ([2d9b50e](https://github.com/kontrollanten/preact-leaflet/commit/2d9b50e866aea90d7cae5fb0f9a525509f76f801))
* **Map:** set default zoom to 6 ([a73fea5](https://github.com/kontrollanten/preact-leaflet/commit/a73fea5d54f4b1735b5fcfbf91ca934a3aab9986))
* **Marker:** add support for markers ([1d57d02](https://github.com/kontrollanten/preact-leaflet/commit/1d57d02a861a07cd9db6f0075b7d3fd505012f5e))
* **MarkerCluster:** add marker cluster support ([6ebe55f](https://github.com/kontrollanten/preact-leaflet/commit/6ebe55f3d407955ed0ed1824130493b39f6dead1))
* **Polyline:** add Polyline support ([8e8f1ce](https://github.com/kontrollanten/preact-leaflet/commit/8e8f1ce9df0d510ad92559fe7153a70427ece271))
* **ZoomControl:** add ZoomControl support ([ae6f788](https://github.com/kontrollanten/preact-leaflet/commit/ae6f7883a733d4af52ae17971cd8252fa364398d))


### Bug Fixes

* make preact X compat ([e96e0ad](https://github.com/kontrollanten/preact-leaflet/commit/e96e0adc660d2d09dd36d1e74d2121cc3edc18f8))
* **deps:** update dependency preact to v10 ([eef9641](https://github.com/kontrollanten/preact-leaflet/commit/eef9641c5a13c5db894c198109e4f0d1fbd6398c))
* **eslint:** v18 config ([f4aa9f0](https://github.com/kontrollanten/preact-leaflet/commit/f4aa9f09a9ffca663677411ce4a5f3b5ae2dc09a))
* add displayName to components ([f58bb26](https://github.com/kontrollanten/preact-leaflet/commit/f58bb2685b557b7e7fa19bd95bf1094f2fe97747))
* **demo:** set accurate title ([82202fc](https://github.com/kontrollanten/preact-leaflet/commit/82202fc39498937443abb3580edf3f4b916ab15b))
* **Layer:** log error instead of throw exception ([3490042](https://github.com/kontrollanten/preact-leaflet/commit/34900428367adc46e623bffa56cbc5b81359d979))
* **leaflet:** use named import from leaflet ([fedbe8d](https://github.com/kontrollanten/preact-leaflet/commit/fedbe8d86385a7169bb4573e133811e9cb5dbf87))
* **Map:** prevent render null children ([d16ec53](https://github.com/kontrollanten/preact-leaflet/commit/d16ec53465aba9835ba0248df92ffc0ea8772c0e))
* **Map:** remove default center value ([0fd9bcd](https://github.com/kontrollanten/preact-leaflet/commit/0fd9bcdb5573dcfc77055c5460106b8405e5123c))


e2e/App.js
import { h, Component, Fragment } from 'preact';
import 'preact/debug';
import { divIcon } from 'leaflet';
import {
  Map, Marker, MarkerCluster, Polyline, TileLayer,
} from '../src';
import 'leaflet/dist/leaflet.css';
import 'leaflet.markercluster/dist/MarkerCluster.Default.css';
import './styles.css';
import route from './long-route';

const getRouteConfig = (hash) => {
  switch (hash) {
    case '#cluster':
      return {
        mapCenter: [63.83919, 20.15069],
        markerCluster: [
          [63.8374896962485, 20.163206074534],
          [63.9374896962485, 20.163206074534],
          [63.7374896962485, 20.163206074534],
          [63.8574896962485, 20.163206074534],
          [63.8174896962485, 20.163206074534],
          [63.8274896962485, 20.163206074534],
          [63.8974896962485, 20.163206074534],
        ],
        markers: [],
        polylines: [],
        zoom: 8,
      };
    case '#polyline':
      return {
        mapCenter: [63.83919, 20.15069],
        markerCluster: false,
        markers: [
          [59.3367, 18.0667],
          [63.8374896962485, 20.163206074534],
        ],
        polylines: [
          route,
        ],
        zoom: 5,
      };
    default:
      return {
        mapCenter: [59.3367, 18.0667],
        markerCluster: false,
        markers: [],
        polylines: [],
        zoom: 10,
      };
  }
};

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = getRouteConfig(window.location.hash);
  }

  componentDidMount() {
    window.addEventListener('popstate', () => {
      this.setState(getRouteConfig(window.location.hash));
    });
  }

  render(props, {
    mapCenter, markerCluster, markers, polylines, zoom,
  }) {
    return (
      <Fragment>
        <header>
          <h1>preact-leaflet</h1>
          <ul className="menu">
            {[
              { link: '#simple', title: 'Simple' },
              { link: '#cluster', title: 'Cluster' },
              { link: '#polyline', title: 'Polyline and markers' },
            ].map(({ link, title }) => (
              <li className={window.location.hash === link ? 'active' : ''}>
                <a href={link}>{title}</a>
              </li>
            ))}
          </ul>
        </header>
        <Map center={mapCenter} style={{ height: '100%' }} zoom={zoom}>
          <TileLayer url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
          {markers.map(position => (
            <Marker icon={divIcon()} position={position} />
          ))}
          {polylines.map(positions => (
            <Polyline positions={positions} />
          ))}
          {markerCluster && (
            <MarkerCluster key="cluster">
              {markerCluster.map(position => (
                <Marker key={position.join(',')} icon={divIcon()} position={position} />
              ))}
            </MarkerCluster>
          )}
        </Map>
      </Fragment>
    );
  }
}
